source/opt/ir_context.h

```diff
@@ -0,0 +1,293 @@
+
+  // Returns a vector of pointers to constant-creation instructions in this
+  // context.
+  inline std::vector<Instruction*> GetConstants();
+  inline std::vector<const Instruction*> GetConstants() const;
+
+  // Iterators for annotation instructions contained in this context.
+  inline Module::inst_iterator annotation_begin();
+  inline Module::inst_iterator annotation_end();
+  inline IteratorRange<Module::inst_iterator> annotations();
+  inline IteratorRange<Module::const_inst_iterator> annotations() const;
+
+  // Iterators for capabilities instructions contained in this module.
+  inline Module::inst_iterator capability_begin();
+  inline Module::inst_iterator capability_end();
+  inline IteratorRange<Module::inst_iterator> capabilities();
+  inline IteratorRange<Module::const_inst_iterator> capabilities() const;
+
+  // Iterators for types, constants and global variables instructions.
+  inline ir::Module::inst_iterator types_values_begin();
+  inline ir::Module::inst_iterator types_values_end();
+  inline IteratorRange<Module::inst_iterator> types_values();
+  inline IteratorRange<Module::const_inst_iterator> types_values() const;
+
+  // Iterators for extension instructions contained in this module.
+  inline Module::inst_iterator ext_inst_import_begin();
+  inline Module::inst_iterator ext_inst_import_end();
+
+  // There are several kinds of debug instructions, according to where they can
+  // appear in the logical layout of a module:
+  //  - Section 7a:  OpString, OpSourceExtension, OpSource, OpSourceContinued
+  //  - Section 7b:  OpName, OpMemberName
+  //  - Section 7c:  OpModuleProcessed
+  //  - Mostly anywhere: OpLine and OpNoLine
+  //
+
+  // Iterators for debug 1 instructions (excluding OpLine & OpNoLine) contained
+  // in this module.  These are for layout section 7a.
+  inline Module::inst_iterator debug1_begin();
+  inline Module::inst_iterator debug1_end();
+  inline IteratorRange<Module::inst_iterator> debugs1();
+  inline IteratorRange<Module::const_inst_iterator> debugs1() const;
+
+  // Iterators for debug 2 instructions (excluding OpLine & OpNoLine) contained
+  // in this module.  These are for layout section 7b.
+  inline Module::inst_iterator debug2_begin();
+  inline Module::inst_iterator debug2_end();
+  inline IteratorRange<Module::inst_iterator> debugs2();
+  inline IteratorRange<Module::const_inst_iterator> debugs2() const;
+
+  // Iterators for debug 3 instructions (excluding OpLine & OpNoLine) contained
+  // in this module.  These are for layout section 7c.
+  inline Module::inst_iterator debug3_begin();
+  inline Module::inst_iterator debug3_end();
+  inline IteratorRange<Module::inst_iterator> debugs3();
+  inline IteratorRange<Module::const_inst_iterator> debugs3() const;
+
+  // Clears all debug instructions (excluding OpLine & OpNoLine).
+  inline void debug_clear();
+
+  // Appends a capability instruction to this module.
+  inline void AddCapability(std::unique_ptr<Instruction> c);
```

Comment:
std::unique_ptr<Instruction>&&?

Reply:
this is actually fine you can move the unique_ptr into the param and then the function will have to move again into the final location.

It's an extra move but for unique_ptr it's just a pointer copy.

Reply:
I don't agree. In one case you "can move", in another you "must move".

Reply:
unique_ptr must be moved anyway

Reply:
That's good of them. So the only functionality we are getting from this is that we do the move twice instead of doing it once?

Reply:
I'll change it just to be more precise.